### PR TITLE
http3: support WebTransport

### LINKED ISSUE #3191: WebTransport Support
Hi, I try to connect http3 server writed by quic-go.
And I setup https by tls certification files.
When I try to use WebTransport in web:
```
let trans = new WebTransport('https://rtc1.live.corp.xxxx.com:443/live/2000.flv');
```

But it fail and console log is:
```
Failed to establish a connection to https://rtc1.live.corp.xxxx.com/live/2000.flv: net::ERR_QUIC_PROTOCOL_ERROR.QUIC_TLS_CERTIFICATE_UNKNOWN (TLS handshake failure (ENCRYPTION_HANDSHAKE) 46: certificate unknown).
```

but I write http3 client by quicgo and use the http3 client to connect http3 server by quic-go, It work fine.

My chrome version is  91.0.4472.77.

Please help.

### PR DESCRIPTION
**Pull request by ydnar**
This PR adds initial support for the WebTransport draft.

To see this in action, open these demos in Chrome Canary:

- Simple demo: https://basic.webtransport.dev/
- Multiplayer mouse cursor demo: https://cursors.webtransport.dev/

### Background

Internally, it exposes a bit of additional API from the `http3` package, including an `http3.Conn` which wraps `quic.EarlySession` and an `http3.RequestStream` which wraps `quic.Stream`.

An `http3.Conn` allows client and server code to share common stream and frame handling, as well as providing a foundation for (de)multiplexing datagrams in support of the [MASQUE datagram draft](https://www.ietf.org/archive/id/draft-ietf-masque-h3-datagram-03.html). An `http3.RequestStream` is used to read and write `HEADERS` and `DATA` frames, as well as process other HTTP/3 frames like MASQUE `CAPSULE` frames.

- An `http3.ServerConn` is constructed using `http3.Accept(quic.EarlySession, Settings)`.
- An `http3.ClientConn` is constructed using `http3.Open(quic.EarlySession, Settings)`.
- A server `http3.RequestStream` is accepted from `(ServerConn).AcceptRequestStream(context.Context)`.
- A client `http3.RequestStream` is opened with `(ClientConn).OpenRequestStream(context.Context)`.

It is possible to construct alternate HTTP/3 clients or servers using these primitives.

This PR builds on #3238 and #3254. It fixes #3191, and replaces #3226. It also adds support for HTTP trailers, modeled after the trailers interface in the standard `net/http` package.

### To do

- [x] Datagram support
- [x] Update existing tests
- [x] Wait for SETTINGS: https://github.com/ietf-wg-webtrans/draft-ietf-webtrans-http3/pull/56
- [x] Trailers support
- [x] WebTransport support in `http3.RoundTripper`
- [x] Tests for WebTransport features
- [ ] Integration tests with headless Chromium

### Screenshots

Chrome demonstrating WebTransport unidirectional, bidirectional streams, and datagrams:

quic-go-webtransport-datagrams-2

### CODE REVIEW

File: http3/client.go [hunk 1]
```diff
@@ -91,111 +88,69 @@ func newClient(
...
-	quicvarint.Write(buf, streamTypeControlStream)
-	// send the SETTINGS frame
-	(&settingsFrame{Datagram: c.opts.EnableDatagram}).Write(buf)
-	_, err = str.Write(buf.Bytes())
-	return err
+
+	go c.handleUnidirectionalStreams()
+
+	return nil
 }
```
Comment (ydnar) [inline http3/client.go:140]: @marten-seemann this is where Open is called for the client.

File: http3/stream.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+func (t StreamType) Valid() bool {
+	return t <= quicvarint.Max
+}
+
+// ReadableStream represents the receiver side of a unidirectional HTTP/3 stream.
+type ReadableStream interface {
+	quic.ReceiveStream
+	Conn() Conn
+	StreamType() StreamType
+}
```
Comment (ydnar) [inline http3/stream.go:49]: I chose `ReadableStream` instead of `ReceiveStream` in part because of this change: https://github.com/w3c/webtransport/issues/306

Second, to better adhere to Go interface naming conventions because a `ReadableStream` implements `io.Reader`.

**Comment by ydnar** [comment]
@marten-seemann could I get your feedback on this PR?

thanks!

File: closed_session.go [hunk 3]
```diff
@@ -89,7 +89,7 @@ func (s *closedLocalSession) destroy(error) {
 	})
 }
 
-func (s *closedLocalSession) getPerspective() protocol.Perspective {
+func (s *closedLocalSession) Perspective() Perspective {
```
Comment (marten-seemann) [inline closed_session.go:92]: I still think this should be handled on the `http3` layer, not on the `quic` layer.

File: http3/conn.go [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
+	acceptUniStream(context.Context, SessionID) (quic.ReceiveStream, error)
+	openStream(SessionID) (quic.Stream, error)
+	openStreamSync(context.Context, SessionID) (quic.Stream, error)
+	openUniStream(SessionID) (quic.SendStream, error)
+	openUniStreamSync(context.Context, SessionID) (quic.SendStream, error)
+	readDatagram(context.Context, SessionID) ([]byte, error)
+	writeDatagram(SessionID, []byte) error
+}
+
+type connection struct {
```
Comment (marten-seemann) [inline http3/conn.go:54]: To elaborate, you can have a `bool` here which tells you if this is a client or a server connection. Then you don't need to ask the `Session` for its `Perspective()`. You just have to set that `bool` once when you initialize this struct.

File: http3/conn.go [hunk 5]
```diff
@@ -0,0 +1,530 @@
...
+	r := quicvarint.NewReader(str)
+	t, err := quicvarint.Read(r)
+	if err != nil {
+		str.CancelRead(quic.StreamErrorCode(errorGeneralProtocolError))
+		return
+	}
+	streamType := StreamType(t)
+
+	// Store control, QPACK, and push streams on conn
+	if streamType < 4 {
```
Comment (marten-seemann) [inline http3/conn.go:283]: Where does this magic number come from?

File: http3/conn.go [hunk 6]
```diff
@@ -0,0 +1,530 @@
...
+	peerStreams      [4]quic.ReceiveStream
+
+	incomingStreamsOnce    sync.Once
+	incomingStreamsErr     error
+	incomingRequestStreams chan RequestStream
+
+	incomingStreamsMutex sync.Mutex
+	incomingStreams      map[SessionID]chan quic.Stream // Lazily constructed
+
+	incomingUniStreamsOnce  sync.Once
```
Comment (marten-seemann) [inline http3/conn.go:73]: Would it make sense to start the go routine right away, and get rid of this `sync.Once`? Every HTTP/3 session will use unidirectional streams.

**Review by marten-seemann — COMMENTED** [review]
Not a complete review yet, but I already left some comments. Do you have an example how a WebTransport application would use this API? That would make it easier for me to understand.

**Comment by ydnar** [comment]
> Not a complete review yet, but I already left some comments. Do you have an example how a WebTransport application would use this API? That would make it easier for me to understand.

Thanks!

I'll throw an example together. Are you OK with an examples directory, or would you prefer a separate repo?

FWIW I'm not happy yet with the WebTransport API here. It's still a bit clunky for the server side.

File: http3/conn.go (hunk 4 above)
Comment (ydnar) [reply http3/conn.go:54]: How would Open or Accept know to error on the wrong perspective if it's not exposed at the quic layer?

File: http3/conn.go (hunk 5 above)
Comment (ydnar) [reply http3/conn.go:283]: 4 standard H3 streams: control, qpack, and push (0x0-0x3).

File: http3/conn.go (hunk 6 above)
Comment (ydnar) [reply http3/conn.go:73]: True.

Deferring it simplifies the test mocking. My thought was that a Conn wouldn't start any goroutines  until asked to minimize resource consumption.

**Comment by marten-seemann** [comment]
> I'll throw an example together. Are you OK with an examples directory, or would you prefer a separate repo?

Since WebTransport will be explicitly supported by the http3 package, it’s justified (and super helpful!) to have an example in this repository.

Thank you for all the effort you’re putting into this!

**Comment by ydnar** [comment]
@marten-seemann I merged my other WIP branch into this one, which reorganizes the HTTP/3 primitives into `Conn`, `RequestStream`, and `FrameWriter`. I was able to easily add support for HTTP trailers with this, and simplify some of the error handling to match the H3 spec.

Next up is building a WebTransport example and tests.

I’d love it if you could take another look, thanks!

**Comment by ydnar** [comment]
@marten-seemann any reason you’re still keeping CircleCI around? Doesn’t GH Actions cover these?

**Comment by ydnar** [comment]
Landed integration tests for WebTransport!

**Comment by ydnar** [comment]
Well that’s odd: https://github.com/lucas-clemente/quic-go/runs/3498414989?check_suite_focus=true#step:6:884

**Comment by ydnar** [comment]
Demo server built on this code: https://basic.webtransport.dev/

**Comment by ydnar** [comment]
Multiplayer mouse cursor demo: https://cursors.webtransport.dev/

File: http3/body.go [hunk 7]
```diff
@@ -1,84 +1,56 @@
...
 	// only set for the http.Response
 	// The channel is closed when the user is done with this response:
 	// either when Read() errors, or when Close() is called.
 	reqDone       chan<- struct{}
 	reqDoneClosed bool
 
-	onFrameError func()
-
-	bytesRemainingInFrame uint64
+	onTrailers trailerFunc
```
Comment (marten-seemann) [inline http3/body.go:22]: Is trailer parsing something we have to do in this PR? It would be really helpful to separate stuff like this into smaller, self-contained PRs.

File: http3/roundtrip.go [hunk 8]
```diff
@@ -41,11 +41,23 @@ type RoundTripper struct {
 	// If nil, reasonable default values will be used.
 	QuicConfig *quic.Config
 
+	// Settings specifies the HTTP/3 settings transmitted to the server.
+	// If nil, reasonable default values will be used.
+	// See https://quicwg.org/base-drafts/draft-ietf-quic-http.html#name-http-2-settings-parameters.
+	// If non-nil, overrides EnableDatagrams and EnableWebTransport.
```
Comment (marten-seemann) [inline http3/roundtrip.go:47]: Why do we need this? I don't like the fact that this overrides other options.
Do we need to export `Settings` at all? Ideally, the user of the library wouldn't have to deal with setting code points at all.

File: http3/stream_type.go [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+	// https://www.ietf.org/archive/id/draft-ietf-quic-qpack-21.html#name-encoder-and-decoder-streams
+	StreamTypeQPACKEncoder StreamType = 0x02
+	StreamTypeQPACKDecoder StreamType = 0x03
+
+	// https://www.ietf.org/archive/id/draft-ietf-webtrans-http3-01.html#section-7.4
+	StreamTypeWebTransportStream StreamType = 0x54
+)
+
+// StreamType represents an HTTP/3 stream type.
+type StreamType uint64
```
Comment (marten-seemann) [inline http3/stream_type.go:23]: Why is this needed? Why is it exported?

File: http3/context_extension_type.go [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+import "fmt"
+
+const (
+	ContextExtensionTypeCloseCode ContextExtensionType = 0x00
+	ContextExtensionTypeDetails   ContextExtensionType = 0x01
+)
+
+// A ContextExtensionType represents an HTTP datagram context extension type code.
+// https://www.ietf.org/archive/id/draft-ietf-masque-h3-datagram-03.html#name-context-extension-types
+type ContextExtensionType uint64
```
Comment (marten-seemann) [inline http3/context_extension_type.go:12]: This type seems entirely unused.

File: http3/grease.go [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+	}
+	return 0x1f*n + 0x21
+}
+
+// writeGreaseFrame writes a greasing frame to w. HTTP/3 peers MUST ignore
+// reserved frame types. The value of n should be somewhat random. The greasing
+// value will be clamped to GreaseMax.
+func writeGreaseFrame(w quicvarint.Writer, n uint64) {
+	quicvarint.Write(w, Grease(n))
+	quicvarint.Write(w, 0) // Zero frame payload length
```
Comment (marten-seemann) [inline http3/grease.go:31]: Better: this should probably write a (short) frame.

File: http3/headers.go [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+	pseudoHeaderScheme    = ":scheme"
+	pseudoHeaderAuthority = ":authority"
+	pseudoHeaderPath      = ":path"
+	pseudoHeaderProtocol  = ":protocol"
+	pseudoHeaderStatus    = ":status"
+)
+
+// RequestHeaders returns valid HTTP/3 header fields for req, or an error if req
+// is malformed.
+func RequestHeaders(req *http.Request) ([]qpack.HeaderField, error) {
```
Comment (marten-seemann) [inline http3/headers.go:28]: Not sure where this is coming from. I assume that you renamed a file and now Git recognizes it as new? It would be helpful if you revert the renaming, so Git can show the actual changes that happened here.

File: http3/request_stream.go [hunk 13]
```diff
@@ -0,0 +1,402 @@
...
+	"github.com/lucas-clemente/quic-go/quicvarint"
+	"github.com/marten-seemann/qpack"
+)
+
+// isClientBidi returns true if id is a client-initiated bidirectional stream ID.
+func isClientBidi(id uint64) bool {
+	return id&0b11 == 0
+}
+
+type requestStream struct {
```
Comment (marten-seemann) [inline http3/request_stream.go:20]: Where does this come from? I assume some renamed file?

File: http3/capsule_type.go [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+const (
+	CapsuleTypeRegisterDatagramContext   CapsuleType = 0x00
+	CapsuleTypeCloseDatagramContext      CapsuleType = 0x01
+	CapsuleTypeDatagram                  CapsuleType = 0x02
+	CapsuleTypeRegisterDatagramNoContext CapsuleType = 0x03
+)
+
+// A CapsuleType represents an HTTP capsule type.
+// https://www.ietf.org/archive/id/draft-ietf-masque-h3-datagram-03.html#name-capsule-types
+type CapsuleType uint64
```
Comment (marten-seemann) [inline http3/capsule_type.go:14]: This doesn't seem to be used anywhere, is it?

File: http3/frame_reader.go [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+
+	"github.com/lucas-clemente/quic-go/quicvarint"
+)
+
+// FrameReader implements a lightweight HTTP/3 frame reader.
+// After setting R, an otherwise zero-value FrameReader is ready to use.
+// If a frame has already been partially read from R, a FrameReader
+// can be initialized by setting Type to the current frame type and N to
+// the number of bytes remaining in the payload.
+type FrameReader struct {
```
Comment (marten-seemann) [inline http3/frame_reader.go:14]: Why is this exported? What was wrong with the `parseNextFrame` function?

File: http3/frame_type.go [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+	// https://www.ietf.org/archive/id/draft-ietf-webtrans-http3-01.html#section-7.3
+	FrameTypeWebTransportStream FrameType = 0x41
+
+	// https://www.ietf.org/archive/id/draft-ietf-masque-h3-datagram-03.html#section-10.1
+	FrameTypeCapsule FrameType = 0xffcab5
+)
+
+// A FrameType represents an HTTP/3 frame type.
+// https://www.ietf.org/archive/id/draft-ietf-quic-http-34.html#name-frame-definitions
+type FrameType uint64
```
Comment (marten-seemann) [inline http3/frame_type.go:24]: Why is this exported?

**Review by marten-seemann — COMMENTED** [review]
Hi @ydnar,
I'm sorry for the late review. I really appreciate all the effort you're putting into is! The reason I'm so slow with reviews here is that I honestly don't know how to review a PR that touch 5k LOC. Going through it commit-by-commit is also not feasible, considering that this branch contains >200 commits.
It would be a lot easier and faster to review if these were separate, independent PRs.

From a quick glance over your PR, I'm not sure if all changes here are actually used and needed to make WebTransport work. For example, it's not clear to me
1. what the CAPSULE frame does. You're defining it, but it doesn't seem to be used (unless I'm missing something)
2. why we're adding HTTP Trailer support in this PR. As far as I understand, it's not needed for WebTransport, is it?

If they're actually not necessary, could you remove them from this PR (along with all the other things that are not needed)? If they are necessary, they sound like changes that could be split into separate PRs.

File: http3/grease.go (hunk 11 above)
Comment (ydnar) [reply http3/grease.go:31]: Good idea!

File: http3/request_stream.go (hunk 13 above)
Comment (ydnar) [reply http3/request_stream.go:20]: The `requestStream` type is entirely new.

File: http3/context_extension_type.go (hunk 10 above)
Comment (ydnar) [reply http3/context_extension_type.go:12]: `CapsuleType` and `ContextExtensionType` are intended to support the [HTTP/3 MASQUE datagram draft](https://www.ietf.org/archive/id/draft-ietf-masque-h3-datagram-03.html), which the WebTransport draft is converging against.

This can be deferred to a future PR.

File: http3/capsule_type.go (hunk 14 above)
Comment (ydnar) [reply http3/capsule_type.go:14]: `CapsuleType` and `ContextExtensionType` are intended to support the [HTTP/3 MASQUE datagram draft](https://www.ietf.org/archive/id/draft-ietf-masque-h3-datagram-03.html), which the WebTransport draft is converging against.

This can be deferred to a future PR.

File: http3/frame_reader.go (hunk 15 above)
Comment (ydnar) [reply http3/frame_reader.go:14]: `FrameReader` allows the caller to determine which frames to skip, and makes no attempt to parse the underlying frames. This was necessary because different stream types (and perspectives) support specific frame types, and the spec(s) require closing the stream or connection in the event of certain frame types.

Comment (ydnar) [reply http3/frame_reader.go:14]: It’s exported so callers implementing lower-level parts of the HTTP/3 protocol could use the same mechanism. In theory, this could go in a `http3guts` package.

File: http3/frame_type.go (hunk 16 above)
Comment (ydnar) [reply http3/frame_type.go:24]: `FrameType` constants are (IETF) spec-defined values which can be used by callers implementing parts of HTTP/3 directly.

File: http3/body.go (hunk 7 above)
Comment (ydnar) [reply http3/body.go:22]: Trailers are not required for WebTransport, but were trivial to add using the new `http3.Conn` and `RequestStream` primitives. I can separate this out.

File: http3/roundtrip.go (hunk 8 above)
Comment (ydnar) [reply http3/roundtrip.go:47]: My preference is to remove `EnableDatagrams` and `EnableWebTransport` in favor of letting the caller specify `Settings`.

The `Settings` type has helpers for enabling datagrams or WebTransport, or other common settings.

File: http3/headers.go (hunk 12 above)
Comment (ydnar) [reply http3/headers.go:28]: It wasn’t so much of a rename as a refactor of `(*requestWriter).writeHeaders`. The rest of `requestWriter` was subsumed by `http3.client` and `http3.RequestStream`.

**Comment by Atul-Bari** [comment]
> @marten-seemann I merged my other WIP branch into this one, which reorganizes the HTTP/3 primitives into `Conn`, `RequestStream`, and `FrameReader`. I was able to easily add support for HTTP trailers with this, and simplify some of the error handling to match the H3 spec.
> 
> Next up is building a WebTransport example and tests.
> 
> I’d love it if you could take another look, thanks!

Hey @ydnar
Can you share an example to test this
I am trying this [https://gist.github.com/ydnar/f73e1806c50c83f90f9b46e13bf751fc](url) for localhost server,

**Comment by ydnar** [comment]
@marten-seemann I updated the description on #3284 to clarify what it is. How can I help you get #3284 reviewed and merged so we can land datagram and WebTransport support?

**Comment by gedw99** [comment]
Wow this looks great 

can’t wait to try this compared to web sockets

**Comment by gedw99** [comment]
@marten-seemann 

is there a possibility that this will be merged ?  

Are there technical blockers that your worried about or is it more just lack of time ?

**Comment by adriancable** [comment]
@ydnar - I am very interested in this. I think that WebTransport support is important for the future of quic-go, and it's sad to see it has not yet been merged.

A couple of questions. @marten-seemann can you share whether this not being merged is due to lack of time or lack of interest? You sharing that with us would be very helpful to the community on how best to move forward.

@ydnar - can you kindly share the disco/DISCO bidirectional stream example client (in other words, what's running at 'the other end' of your browser screenshot) to make this easier to test? Thank you.

**Comment by ydnar** [comment]
> @ydnar - can you kindly share the disco/DISCO bidirectional stream example server (in other words, what's running at 'the other end' of your browser screenshot) to make this easier to test? Thank you.

Sure. I need to dust off the branch that enables this. I’d prefer not to maintain a fork of quic-go just to support WebTransport (or a fork of quic/http3).

There’s some relevant stuff in Go 1.18 that should improve QUIC performance too (allocation-free UDP, net/netip package).

**Comment by adriancable** [comment]
Hi @ydnar - actually, I got it figured out.

I decided to develop a new Go package which implements a WebTransport server over HTTP/3 (updated per the latest IETF drafts). Its API is inspired by the PR’s API but it implements it with the standard quic-go package as a dependency. It has pretty complete support for all of WebTransport including datagrams, server and client initiated uni/bidi streams and so on. It’s currently fully functional but needs docs written and further testing. I’ll share the link when ready if interested.

**Comment by ydnar** [comment]
> Hi @ydnar - actually, I got it figured out.
> 
> I decided to develop a new Go package which implements a WebTransport server over HTTP/3 (updated per the latest IETF drafts). Its API is inspired by the PR’s API but it implements it with the standard quic-go package as a dependency. It has pretty complete support for all of WebTransport including datagrams, server and client initiated uni/bidi streams and so on. It’s currently fully functional but needs docs written and further testing. I’ll share the link when ready if interested.

Sure, I’ll take a look!

**Comment by adriancable** [comment]
> Sure, I’ll take a look!

@ydnar - great!

Package is here: [github.com/adriancable/webtransport-go](https://github.com/adriancable/webtransport-go)

Example server & client are here: [github.com/adriancable/webtransport-go-example](https://github.com/adriancable/webtransport-go-example)

GoDoc is here: [pkg.go.dev/github.com/adriancable/webtransport-go](https://pkg.go.dev/github.com/adriancable/webtransport-go)

Hope that someone finds this useful.

**Comment by marten-seemann** [comment]
Closing, since there's now a WebTransport implementation using quic-go: https://github.com/marten-seemann/webtransport-go.
